### PR TITLE
 JAMES-2763 ES startup check

### PR DESCRIPTION
**Pull request by trantienduchn**
Not done yet, still missing the adding configuration item for supporting version

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchRestAPITest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+class ElasticSearchRestAPITest {
+
+    @RegisterExtension
+    static DockerElasticSearchExtension linshareExtension = DockerElasticSearchExtension.withSingletonES();
+
+    @Test
+    void aboutShouldReturnTheRightVersion() {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchRestAPITest.java]: why don't you check all properties?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            return name;
+        }
+
+        public Version getVersion() {
+            return version;
+        }
+    }
+
+    @VisibleForTesting
+    static ElasticSearchRestAPI from(Host esHost) throws RetryableException {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java]: please don't leak feign specific exceptions

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 3]
```diff
@@ -37,23 +37,41 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(ElasticSearchStartUpCheck.class);
     private static final String CHECK_NAME = "ElasticSearchStartUpCheck";
     private static final String CAN_NOT_FIND_VERSION_MESSAGE = "cannot find ElasticSearch version";
+    private static final String SUGGESTION_VERSION_IS_NOT_MATCHED_MESSAGE = "suggested ElasticSearch version (%s) is not match " +
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: s/is not matched/does not match

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheckTest.java [hunk 4]
```diff
@@ -51,4 +61,16 @@ void checkShouldReturnBadWhenUnAbleToConnectToElasticSearch() {
         assertThat(testee.check().isBad())
             .isTrue();
     }
+
+    @Test
+    void checkShouldReturnBadWhenESVersionIsNotMatchedWithTheSuggestion() {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheckTest.java]: doesNotMatch

**Comment by trantienduchn** [comment]
```
[ea4be3475fe233a241c21fff5016a2c0f8b6e7d3] [INFO] There is 1 error reported by Checkstyle 8.18 with checkstyle.xml ruleset.
[ea4be3475fe233a241c21fff5016a2c0f8b6e7d3] [ERROR] src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java:[43,55] (whitespace) WhitespaceAround: WhitespaceAround: '=' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)
```

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java (hunk 2 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java]: We have a cluster, not a single host. What if server1 is version 2.2 and server2 6.6? If we only check the right server then we are screwed.

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ ****************************************************************/
+
+package org.apache.james.modules.mailbox;
+
+import org.apache.james.util.Host;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+
+import feign.Feign;
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchRestAPI.java]: Beware of  hammer-nail syndrom...

Feign is a beautiful, effective tool, but better ways of acheiving our goals exists:

```
        Arrays.stream(client.admin().cluster().clusterStats(new ClusterStatsRequest()).get().getNodes())
            .filter(node -> !isValid(node.nodeInfo().getVersion()))
            ...;
```

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        this.esHttpHost = esHttpHost;
+    }
+
+    @Override
+    public CheckResult check() {
+        try {
+            About.Version version = getVersion();
+        } catch (RetryableException e) {
+            String errorMessage = CAN_NOT_FIND_VERSION_MESSAGE + ": " + e.getMessage();
+            LOGGER.error(errorMessage);
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: Are we loosing a stacktrace?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    @VisibleForTesting
+    ElasticSearchStartUpCheck(Host esHttpHost) {
+        this.esHttpHost = esHttpHost;
+    }
+
+    @Override
+    public CheckResult check() {
+        try {
+            About.Version version = getVersion();
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: We don't do an assertion like `Version should be at least 2.2` ?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        } catch (RetryableException e) {
+            String errorMessage = CAN_NOT_FIND_VERSION_MESSAGE + ": " + e.getMessage();
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: `CAN_NOT_FIND_VERSION_MESSAGE` -> How do you know?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+public class ElasticSearchStartUpCheck implements StartUpChecksPerformer.StartUpCheck {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ElasticSearchStartUpCheck.class);
+    private static final String CHECK_NAME = "ElasticSearchStartUpCheck";
+    private static final String CAN_NOT_FIND_VERSION_MESSAGE = "cannot find ElasticSearch version";
+
+    private final Host esHttpHost;
+
+    @Inject
+    public ElasticSearchStartUpCheck(ESReporterConfiguration esConfiguration) {
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: Are you specifically checking the ElasticSearch instance used for metric collection?

We should concentrate on the one providing the search feature - they could be different...

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheckTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+package org.apache.james.modules.mailbox;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.apache.james.backends.es.DockerElasticSearchExtension;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+class ElasticSearchStartUpCheckTest {
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheckTest.java]: We should have a test with an ElasticSearch of the wrong version

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java [hunk 11]
```diff
@@ -37,23 +37,41 @@
...
         this.esHttpHost = esHttpHost;
+        this.esVersionSuggestion = esVersionSuggestion;
     }
 
+
     @Override
     public CheckResult check() {
         try {
             About.Version version = getVersion();
+            if (!version.getNumber().equals(ES_VERSION_SUGGESTION)) {
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: Mmmh I am not sure we should be too restrictive about this...

Actually we accept 2.4.x so starting with a 2.4.7 should be OK. Maybe even from a 2.5.2 but 3.x should be rejected.

I believe we need to parse version number into business objects which we can support advanced comparaisons on...

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java (hunk 9 above)
Comment (trantienduchn) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchStartUpCheck.java]: ESReporterConfiguration is the only instance carrying the ES http host, I want to use it temporally to have a tests suite for this class. Then I will think about the alternative instance

**Comment by chibenwa** [comment]
> Cannot you rely on an ElasticSearch driver for verifying version compatibility?

Yes following the chat discussion we had this seems reasonable.

We can still rely on Startup checks for other places where it makes sens.

**Comment by trantienduchn** [comment]
For the record: ES driver doesn't work well when connecting to ES servers in version the driver does not compatible. It throws NoNodeAvailableException, using a ES server with a not compatible version probably makes james startup fails and shutdown james.

Stop the work on this

**Comment by chibenwa** [comment]
> For the record: ES driver doesn't work well when connecting to ES servers in version the driver does not compatible.

We need an integration test on this behaviour this should be enough for now.

>  It throws NoNodeAvailableException, using a ES server with a not compatible version probably makes james startup fails and shutdown james.

We can legitimately expect ES 6.x being more versatile on this.

Let's not invest too much on ES 2.x not-good-enough exception message.

**Comment by Arsnael** [comment]
Read it

**Comment by chibenwa** [comment]
As explained before, this work is not needed as we can rely on the driver for these checks.

**Pull request closed without merging**